core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/sync/MixedPackageVersionService.java

```diff
@@ -0,0 +1,249 @@
+package com.sequenceiq.cloudbreak.service.upgrade.sync;
+
+import static com.sequenceiq.cloudbreak.api.endpoint.v4.common.Status.UPDATE_IN_PROGRESS;
+import static com.sequenceiq.cloudbreak.event.ResourceEvent.STACK_CM_MIXED_PACKAGE_VERSIONS_FAILED;
+import static com.sequenceiq.cloudbreak.event.ResourceEvent.STACK_CM_MIXED_PACKAGE_VERSIONS_FAILED_MULTIPLE;
+import static com.sequenceiq.cloudbreak.event.ResourceEvent.STACK_CM_MIXED_PACKAGE_VERSIONS_FAILED_NO_CANDIDATE;
+import static com.sequenceiq.cloudbreak.event.ResourceEvent.STACK_CM_MIXED_PACKAGE_VERSIONS_NEWER_FAILED;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.cloud.VersionComparator;
+import com.sequenceiq.cloudbreak.cloud.model.ClouderaManagerProduct;
+import com.sequenceiq.cloudbreak.cloud.model.catalog.Image;
+import com.sequenceiq.cloudbreak.core.CloudbreakImageCatalogException;
+import com.sequenceiq.cloudbreak.core.CloudbreakImageNotFoundException;
+import com.sequenceiq.cloudbreak.event.ResourceEvent;
+import com.sequenceiq.cloudbreak.service.image.ClusterUpgradeTargetImageService;
+import com.sequenceiq.cloudbreak.service.image.ImageCatalogService;
+import com.sequenceiq.cloudbreak.service.image.ImageService;
+import com.sequenceiq.cloudbreak.service.image.StatedImage;
+import com.sequenceiq.cloudbreak.service.parcel.ClouderaManagerProductTransformer;
+import com.sequenceiq.cloudbreak.service.upgrade.sync.common.ParcelInfo;
+import com.sequenceiq.cloudbreak.service.upgrade.sync.operationresult.CmSyncOperationResult;
+import com.sequenceiq.cloudbreak.structuredevent.event.CloudbreakEventService;
+
+@Service
+public class MixedPackageVersionService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MixedPackageVersionService.class);
+
+    private static final String CM_KEY = "cm";
+
+    private static final String CLOUDERA_MANAGER_NAME = "Cloudera Manager";
+
+    @Inject
+    private CloudbreakEventService eventService;
+
+    @Inject
+    private ClusterUpgradeTargetImageService clusterUpgradeTargetImageService;
+
+    @Inject
+    private ClouderaManagerProductTransformer clouderaManagerProductTransformer;
+
+    @Inject
+    private ImageService imageService;
+
+    @Inject
+    private ImageCatalogService imageCatalogService;
+
+    void validatePackageVersions(Long stackId, CmSyncOperationResult syncResult, Set<Image> candidateImages) {
+        Optional<String> activeCmVersion = syncResult.getCmRepoSyncOperationResult().getInstalledCmVersion();
+        if (activeCmVersion.isPresent()) {
```

Comment:
I wonder if it makes sense to do the validation if cm version is present but for some reason there is no parcel info from the server. Not clear, however, how can we have CM version but no parcel versions.
There is a task in phase 2 to read the CM version from yum if CM would not respond. If that get's implemented then it might cause such a situation that only CM version is present.
It might make sense to check CM version alone, but then messages might need to be changed (e.g. "... is not a tested combination" would not make much sense)